File: src/ffi/uvl.c
#include <stddef.h>
#include <stub_ffi.h>
#include <psp2/uvl.h>

void ffi_register_uvl()
{
	static Function funcs[] =
	{
		{"uvl_alloc_code_mem", uvl_alloc_code_mem},
		{"uvl_unlock_mem", uvl_unlock_mem},
		{"uvl_lock_mem", uvl_lock_mem},
		{"uvl_flush_icache", uvl_flush_icache},
		{"uvl_load", uvl_load},
		{"uvl_exit", uvl_exit},
		{"uvl_log_write", uvl_log_write},
		{NULL, NULL}
	};
	static FunctionTable table = { .funcs = funcs, .next = NULL };
	ffi_add_table(&table);
}

File: lua/uvl.lua
-- misc
local ffi = require 'ffi'

ffi.cdef [[
void *uvl_alloc_code_mem(unsigned int *p_len);
void uvl_unlock_mem(void);
void uvl_lock_mem(void);
void uvl_flush_icache(void *addr, unsigned int len);
int uvl_load(const char *path);
void uvl_exit(int status);
int uvl_log_write(const void* buffer, unsigned int size);
]]

uvl = {}

function uvl.load(path)
  return ffi.C.uvl_load(path) == 0
end

function uvl.exit(status)
  vita2d.fini() -- Some cleanup.
  physfs.deinit()

  ffi.C.uvl_exit(tonumber(status))
end

os.exit = uvl.exit -- Alias it to os.exit
